### ISSUE TITLE
Refactor `@wordpress/components` package to TypeScript

### ISSUE THREAD

**Issue by andrewserong** [WordPress/gutenberg#35744]
Related to #30503, #28399 and #27484

### Context ✨

We would like to refactor the entire `@wordpress/components` package to TypeScript.

Some components are already written in TypeScript, while the majority of the code is still written in JavaScript.

By refactoring the whole package to TypeScript, we would be able to take advantage of type safety, while also providing first-party types to the package's consumers (instead of the [third-party ones](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/wordpress__components)).

A fully-typed set of components would also potentially allow us to generate documentation programmatically based off types.

### Details of the refactor 🔍 

The refactors should introduce the least amount of runtime changes possible — ideally none. All JS files should be refactored to TypeScript.

You can also refer to [the **TypeScript migration guide** in the `@wordpress/components` package guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript)., and reference an existing component like [`ItemGroup`](https://github.com/wordpress/gutenberg/blob/HEAD/packages/components/src/item-group) or [`ToolsPanel`](https://github.com/wordpress/gutenberg/blob/HEAD/packages/components/src/tools-panel/tools-panel).

### Next up

Once the refactor to TypeScript is complete, we'll be able to:

- migrate all components to Emotion for styles
- refactor folder structures, potentially split into hook+component
- refactor components to use context
- remove (or mark as deprecated) the third party types

### List ⚒️

The components in the [`exclude` list of our tsconfig](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/tsconfig.json) are particularly high priority.

- [x] `AlignmentMatrixControl` (`packages/components/src/alignment-matrix-control`) #46162
- [x] `AnglePickerControl` (`packages/components/src/angle-picker-control`) @ciampo #45820
- [x] `Animate` (`packages/components/src/animate`) #49243
- [x] `Animation` (`packages/components/src/animation`)
- [x] `Autocomplete` (`packages/components/src/autocomplete`) #47751
- [x] `BaseControl` (`packages/components/src/base-control`) #39468
- [x] `BaseField` (`packages/components/src/base-field`) #45712
- [x] `BorderBoxControl` (`packages/components/src/border-box-control`) (Storybook #45002, tests #47755)
- [x] `BorderControl` (`packages/components/src/border-control`) #41843
- [x] `BoxControl` (`packages/components/src/box-control`) #47622
- [x] `Button` (`packages/components/src/button`) #46997
  - [x] update types for components depending on button, like `NavigatorBackButton` and `NavigatorButton`) #47754
- [x] `ButtonGroup` (`packages/components/src/button-group`) #41007
- [x] `Card` (`packages/components/src/card`) #42941
- [x] `CheckboxControl` (`packages/components/src/checkbox-control`) #40915
- [x] `CircularOptionPicker` (`packages/components/src/circular-option-picker`) #47937
- [x] `ClipboardButton` (`packages/components/src/clipboard-button`) #51334
- [x] `ColorIndicator` (`packages/components/src/color-indicator`) #41587
- [x] `ColorListPicker` (`packages/components/src/color-list-picker`) #46358
- [x] `ColorPalette` (`packages/components/src/color-palette`) #44632
- [x] `ColorPicker` (`packages/components/src/color-picker`) @chad1008 #49214
- [x] `ComboboxControl` (`packages/components/src/combobox-control`) #47581
- [x] `Composite` (`packages/components/src/composite`) #54028
- [x] `ConfirmDialog` (`packages/components/src/confirm-dialog`) #54954
- [x] `CustomGradientPicker` (`packages/components/src/custom-gradient-picker`) #48929
- [ ] `CustomSelectControl` (`packages/components/src/custom-select-control`) (instead of a refactor, the component will be rewritten from scratch https://github.com/WordPress/gutenberg/pull/41726)
- [x] `Dashicon` (`packages/components/src/dashicon`) #45924
- [x] `DateTime` (`packages/components/src/date-time`) #40775 #40986 #40957
- [x] `DimensionControl` (`packages/components/src/dimension-control`) #47351
- [x] `Disabled` (`packages/components/src/disabled`) #42708
- [x] `Disclosure` (`packages/components/src/disclosure`) (will be migrated as part of the [reakit -> ariakit refactor](https://github.com/WordPress/gutenberg/issues/53278)) #55639
- [x] `Divider` (`packages/components/src/divider`) #41991
- [x] `Draggable` (`packages/components/src/draggable`) #45471
- [x] `DropZone` (`packages/components/src/drop-zone`) #43962
- [x] `Dropdown` (`packages/components/src/dropdown`) (should also replace types in `ColorPalette`) #45787
- [x] `DropdownMenu` (`packages/components/src/dropdown-menu`)  (suggested [in this comment](https://github.com/WordPress/gutenberg/pull/35415#pullrequestreview-779635786)) @chad1008 https://github.com/WordPress/gutenberg/pull/50187
- [x] `DuotonePicker`, `DuotoneSwatch` (`packages/components/src/duotone-picker`) #49060 
- [x] `Elevation` (`packages/components/src/elevation`) #42302
- [x] `ExternalLink` (`packages/components/src/external-link`) #41681
- [x] `Flex` (`packages/components/src/flex`) #42537
- [x] `FocalPointPicker` (`packages/components/src/focal-point-picker`) #43872
- [x] `FocusableIframe` (`packages/components/src/focusable-iframe`) #53979
- [x] `FontSizePicker` (`packages/components/src/font-size-picker`) #44449
- [x] `FormFileUpload` (`packages/components/src/form-file-upload`) #43960
- [x] `FormToggle` (`packages/components/src/form-toggle`) #41729
- [x] `FormTokenField` (`packages/components/src/form-token-field`) #41216, #43442
- [x] `GradientPicker` (`packages/components/src/gradient-picker`) #48316 @chad1008 
- [x] `Grid` (`packages/components/src/grid`) #41923
- [x] `Guide` (`packages/components/src/guide`) #47493
- [x] `HStack` `packages/components/src/h-stack`) #41580
- [x] `Heading` (`packages/components/src/heading`) #41921
- [x] `HigherOrder` (`packages/components/src/higher-order`)
  - [x] `navigate-regions` #48632
  - [x] `with-constrained-tabbing` #48162
  - [x] `with-fallback-styles` #48720
  - [x] `with-filters` #48721
  - [x] `with-focus-outside` #53980 (also remove the `@ts-expect-error` calls in `ComboboxControl`)
  - [x] `with-focus-return` #48748
  - [x] `with-notices` #49088
  - [x] `with-spoken-messages` #48163
- [x] `Icon` (`packages/components/src/icon`) - convert tests #49066
- [x] `InputControl` (`packages/components/src/input-control`)  #33696
- [x] `IsolatedEventContainer` #54316 (`packages/components/src/isolated-event-container`)
- [x] `ItemGroup` (`packages/components/src/item-group`) (#46945)
- [x] `KeyboardShortcuts` (`packages/components/src/keyboard-shortcuts`) #47429
- [x] `MenuGroup` (`packages/components/src/menu-group`) #36561 #45617
- [x] `MenuItem` (`packages/components/src/menu-item`)  #53132 #37020
- [x] `MenuItemsChoice` (`packages/components/src/menu-items-choice`) #47180
- [x] `Modal` (`packages/components/src/modal`) #42949
- [x] `NavigableContainer` (`packages/components/src/navigable-container`) #49377
- [x] `Navigation` (`packages/components/src/navigation`) #48742
- [x] `Navigator` (`packages/components/src/navigator`) (unit tests: #44970, storybook: #44979)
- [x] `Notice` (`packages/components/src/notice`) #47118
- [ ] `NumberControl` (`packages/components/src/number-control`) (suggested [in this comment](https://github.com/WordPress/gutenberg/issues/34817#issuecomment-941161028)) (include also changes proposed in https://github.com/WordPress/gutenberg/pull/37164#issuecomment-1031959395)
  - [x] #43791
  - [ ] Type the `value` prop and remove the last `@ts-expect-error` comments
    - [x] #45990
    - [ ] #45982
- [x] `PaletteEdit` (`packages/components/src/palette-edit`): #47764 
- [x] `Panel` (`packages/components/src/panel`) #47259
- [x] `Placeholder` (`packages/components/src/placeholder`) #42990 #43069
- [x] `Popover` (`packages/components/src/popover`) #43823 #44373
- [x] `QueryControl` (`packages/components/src/query-controls`) #46721
- [x] ~`Radio` (`packages/components/src/radio`)~
- [x] ~`RadioContext` (`packages/components/src/radio-context`)~
- [x] `RadioControl` (`packages/components/src/radio-control`) #41568
- [x] ~`RadioGroup` (`packages/components/src/radio-group`)~
- [x] `RangeControl` (`packages/components/src/range-control`) (suggested [in this comment](https://github.com/WordPress/gutenberg/issues/34817#issuecomment-941161028)) https://github.com/WordPress/gutenberg/pull/40535, #41444, #41445
- [x] `ResizableBox` (`packages/components/src/resizable-box`) (#46774, #47756)
- [x] `ResponsiveWrapper` (`packages/components/src/responsive-wrapper`) #46480
- [x] `Sandbox` (`packages/components/src/sandbox`) #46478
- [x] `ScrollLock` (`packages/components/src/scroll-lock`) #42303
- [x] `Scrollable` (`packages/components/src/scrollable`) #42016
- [x] `SearchControl` (`packages/components/src/search-control`) #43871
- [x] `SelectControl` (`packages/components/src/select-control`) #40737
  - [x] Improve type of `value` prop (`string[]` if `multiple` is `true`, `string` otherwise) https://github.com/WordPress/gutenberg/pull/47390
- [x] `Shortcut` (`packages/components/src/shortcut`) #42272
- [x] `SlotFill` (`packages/components/src/slot-fill`) #51350
- [x] `Snackbar` (`packages/components/src/snackbar`) (#45472)
- [x] `Spacer` (`packages/components/src/spacer`) #42013
- [x] `Spinner` (`packages/components/src/spinner`) #41540 
- [x] `StyleProvider` (`packages/components/src/style-provider`) #42541
- [x] `Surface` (`packages/components/src/surface`) #41212
- [x] `Swatch` (`packages/components/src/swatch`) #42162
- [x] `TabPanel` (`packages/components/src/tab-panel`) #43536
- [x] `Text` (`packages/components/src/text`) (check truncate behaviors noted in this [comment](https://github.com/WordPress/gutenberg/pull/38842#pullrequestreview-888945388), try [removing boolean types](https://github.com/WordPress/gutenberg/pull/38842#discussion_r816895531) from `adjustLineHeightForInnerControls`) #54953
- [x] `TextControl` (`packages/components/src/text-control`) #40633
- [x] `TextHighlight` (`packages/components/src/text-highlight`) #41698 
- [x] `TextareaControl` (`packages/components/src/textarea-control`) #41215
- [x] `Tip` (`packages/components/src/tip`) #42262
- [x] `ToggleControl` (`packages/components/src/toggle-control`) #43717
- [x] `ToggleGroupControl` (`packages/components/src/toggle-group-control`)
- [x] `Toolbar`:
  - [x] `Toolbar` (`packages/components/src/toolbar`) #47087
  - [x] `ToolbarButton` (`packages/components/src/toolbar-button`) #47750
  - [x] `ToolbarContext` (`packages/components/src/toolbar-context`) #49002
  - [x] `ToolbarDropdownMenu` (`packages/components/src/toolbar-dropdown-menu`) #54321
  - [x] `ToolbarGroup` (`packages/components/src/toolbar-group`) #54317
  - [x] `ToolbarItem` (`packages/components/src/toolbar-item`) #49190
- [x] `ToolsPanel` (`packages/components/src/tools-panel`) @ciampo (Storybook: #47944, Unit tests: #48275)
- [x] `Tooltip` (`packages/components/src/tooltip`) (As per #42753, the component won't be refactored, but will be instead completely rewritten) #48440
- [x] `TreeGrid` (`packages/components/src/tree-grid`) #47516
- [x] `TreeSelect` (`packages/components/src/tree-select`) #41536
- [x] `Truncate` (`packages/components/src/truncate`) #41697
- [x] `UnitControl` (`packages/components/src/unit-control`)  (refactored to TS in [35281](https://github.com/WordPress/gutenberg/pull/35281), suggested follow ups in [this comment](https://github.com/WordPress/gutenberg/pull/35281#pullrequestreview-778443434), tests refactored in #40697)
- [x] `VStack` (`packages/components/src/v-stack`) #41850
- [x] `View` (`packages/components/src/view`) https://github.com/WordPress/gutenberg/pull/45667
- [x] `VisuallyHidden` (`packages/components/src/visually-hidden`) #42220
- [x] `ZStack` (`packages/components/src/z-stack`) #41694
- [x] `index.js`

<details>
<summary>Click to reveal the dependency graph</summary>

![dependencygraph](https://user-images.githubusercontent.com/1083581/150875105-2f977303-ec51-4b45-aeae-e49d0c3418b8.svg)
</details>

**Comment by rajnish93** [comment]
Hello everyone :wave: I want to work on this issue. Can anyone tell me if I need to convert all js file to typescript. (example in **DropdownMenu** have `index.js` , `index.native.js`, `test folder` and `stories` folder) Do I need to change only `index.js` to typescript or all files.

**Comment by ciampo** [comment]
> Hello everyone 👋 I want to work on this issue. Can anyone tell me if I need to convert all js file to typescript. (example in **DropdownMenu** have `index.js` , `index.native.js`, `test folder` and `stories` folder) Do I need to change only `index.js` to typescript or all files.

Hey @rajnish93 , thank you for your interest! I would ask you to wait before contributing to this issue, as I'm going to update the list of components that need refactoring. Some of the refactors are quite complex and it's probably better if we champion a few of them first to understand any caveat along the way.

We'd still love to get your help further down the line though, if that's ok!

**Comment by ciampo** [comment]
Alright, I've edited this issue's description and added all components/folders in the package.

There's still a couple of things of which I’m not sure yet:

1. so far, even with fully-typed components, we left a few files still in JS — in particular unit tests, storybook examples, and react native files. Is there a particular reason for each of those?
2. Where to draw the line in the refactor? On one hand, as we refactor, it would good to hook components to the context system, and therefore use [the `WordPressComponent` type](https://github.com/WordPress/gutenberg/blob/8aa94cc17b774a239430d12ab3aeaab1aa86519f/packages/components/src/ui/context/wordpress-component.ts). On the other hand, we may want to keep the refactor as lean as possible and only aim changing file extension, fixing TS errors, and (maybe) adding a `types.ts` file. We should also consider that certain components are still written in class form — in that case, should we first refactor them to functional components in a separate PR preceding the TypeScript refactor?

Would love to hear folks' opinions on this (cc @mirka @diegohaz @sarayourfriend )

**Comment by sarayourfriend** [comment]
> so far, even with fully-typed components, we left a few files still in JS — in particular unit tests, storybook examples, and react native files. Is there a particular reason for each of those?

From a "value" perspective I don't think there's a compelling reason to spend contributor time on updating unit tests and storybook examples to use TypeScript.

React native is a separate beast altogether. I'm not sure what it would take to convert those to TypeScript or if it's even possible with the current tooling.

> Where to draw the line in the refactor? On one hand, as we refactor, it would good to hook components to the context system, and therefore use the WordPressComponent type. On the other hand, we may want to keep the refactor as lean as possible and only aim changing file extension, fixing TS errors, and (maybe) adding a types.ts file. We should also consider that certain components are still written in class form — in that case, should we first refactor them to functional components in a separate PR preceding the TypeScript refactor?

From a perspective of what will benefit the consumers of the components package the most, just adding types is a big win. The rest of the things help bring the components package inline with itself (context, Emotion, hooks, etc) but don't necessarily have a big effect on the _consumers_ of the package. While converting to TypeScript will increase the maintainability of the package, it probably has the greatest effect on consumers of it. Given it will be the largest most widely used package to be typed in Gutenberg, it will also serve to evangelize the benefits of TypeScript to other contributors, many of whom may still have doubts that TypeScript is a valuable tool to learn in addition to everything else they've already been asked to learn to be able to contribute to Gutenberg (JavaScript, React (twice over now given the transition to hooks happened during Gutenberg's lifetime), etc).

Remember, we [don't refactor just because we can](https://make.wordpress.org/core/handbook/contribute/code-refactoring/). In the spirit of this, we should continue only converting to TypeScript (instead of just annotating with JSDoc) when necessary. This was the [agreement made in the Make post that added TypeScript to Gutenberg](https://make.wordpress.org/core/2021/04/09/follow-up-to-the-native-typescript-proposal/) (see `When to use TypeScript for existing code` section).

**Comment by ciampo** [comment]
> From a "value" perspective I don't think there's a compelling reason to spend contributor time on updating unit tests and storybook examples to use TypeScript.

I see your point, although I liked the idea that a component would have all of its files in TypeScript, rather than a mix of TypeScript and JavaScript. But definitely not a priority when considering which order we should tackle components.

> React native is a separate beast altogether. I'm not sure what it would take to convert those to TypeScript or if it's even possible with the current tooling.

I'll be seeking more advice from mobile folks on this matter specifically and come back with updates

> Remember, we don't refactor just because we can. In the spirit of this, we should continue only converting to TypeScript (instead of just annotating with JSDoc) when necessary. 

Thank you for the link! I would argue that a refactor to TypeScript has become, for this package, quite a necessity. As you explained too, it will bring benefits to the consumers of the package by adding first-party types, allowing us to deprecate (and stop maintaining) the [`DefinitelyTyped`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/wordpress__components) ones

Another added benefit that you mentioned is maintainability: the package is currently very fragmented in terms of technology:
- some components are written in JavaScript, some in TypeScript
- some components are styles in Sass, some in Emotion
- some components use the context system, most don't

Aligning the components in the package on the same tech stack as much as possible would make it easier for devs to maintain the package and for new folks to contribute to it. 

Having said that, in the context of this TypeScript refactor, I think it makes the most sense to keep the refactor as "lean" as possible — i.e. adding the necessary types and refactoring the code without (theoretically) introducing any runtime changes. We can only iterate later on in case we want to introduce more changes to each component when necessary.

**Comment by sarayourfriend** [comment]
> Having said that, in the context of this TypeScript refactor, I think it makes the most sense to keep the refactor as "lean" as possible — i.e. adding the necessary types and refactoring the code without (theoretically) introducing any runtime changes. We can only iterate later on in case we want to introduce more changes to each component when necessary.

Yes! I think this is the clearest path forward for this issue, otherwise the scope is far too large!

**Comment by dcalhoun** [comment]
> so far, even with fully-typed components, we left a few files still in JS — in particular unit tests, storybook examples, and react native files. Is there a particular reason for each of those?

> React native is a separate beast altogether. I'm not sure what it would take to convert those to TypeScript or if it's even possible with the current tooling.

Hi. 👋🏻 I spoke briefly with @ciampo about native modules in a chat, but I wanted to share thoughts here as well. Thanks for the ping on this topic! 🙇🏻 

**TL;DR:** Native modules _can_ import shared/web TypeScript modules successfully, both the ones that exist today and any modules converted in the future. Native modules themselves _cannot_ be easily converted to TypeScript themselves currently due to https://github.com/microsoft/TypeScript/issues/21926. 

---

I do believe the native mobile Gutenberg contributors are interested in typed JavaScript for the project. The project would likely gain a lot of benefit from static analysis and code completion. Also, leveraging [TypeScript with React Native](https://reactnative.dev/docs/typescript) is fairly common in the community at large. 

That said, I researched the possibility of using TypeScript for native modules in this repository mid last year. Unfortunately, it does not appear possible due to https://github.com/microsoft/TypeScript/issues/21926, which outlines TypeScript tooling's lack of support for [platform-specific files](https://reactnative.dev/docs/platform-specific-code#platform-specific-extensions).  

## Current Status of TypeScript for Native Gutenberg Modules

* 🟢 Native modules _do_ successfully import and compile shared TypeScript modules, presumably due to Gutenberg’s [existing TypeScript configuration](https://github.com/WordPress/gutenberg/blob/d8ab9a1db6ef18c38b89ed2790b4d48265740e34/tsconfig.json) and [React Native’s default TypeScript support](https://reactnative.dev/docs/typescript#adding-typescript-to-an-existing-project). E.g. [`MissingEdit`](https://github.com/WordPress/gutenberg/blob/d8ab9a1db6ef18c38b89ed2790b4d48265740e34/packages/block-library/src/missing/edit.native.js#L19) imports [`Icon`](https://github.com/WordPress/gutenberg/blob/d8ab9a1db6ef18c38b89ed2790b4d48265740e34/packages/components/src/icon/index.tsx) from `@wordpress/components`. 
* 🟢 Native modules _do_ currently benefit from code completion of shared modules which include a single, unified interface for both web and native — this applies to both TypeScript and JSDoc definitions. 
* 🔴 Native modules _are not_ statically typed checked via CI or editors due to Gutenberg’s existing TypeScript configuration [excluding native files](https://github.com/WordPress/gutenberg/blob/d8ab9a1db6ef18c38b89ed2790b4d48265740e34/tsconfig.base.json#L37-L39). 
* 🔴 Native modules _cannot_ be statically typed checked via CI or editors as there is no way to configure TypeScript to load [platform-specific files](https://github.com/microsoft/TypeScript/issues/21926) or [enforce a common interface](https://github.com/microsoft/TypeScript/issues/8328#issuecomment-219254627). 
* 🔴 Native modules _cannot_ benefit from code completion from TypeScript and JSDoc definitions _in native modules_, e.g. `index.native.[tsx|js]`. The definitions from the sibling web module is always loaded, as there is no way to configure TypeScript to load [platform-specific files](https://github.com/microsoft/TypeScript/issues/21926). 

While a native module _can technically_ be migrated to a TypeScript file, there is likely little value in a native module TypeScript file if (1) there is no way to enforce adherence to the native TypeScript module interface and (2) its interface cannot diverge from the sibling web TypeScript module. 

So, with all of that, my personal opinion is that it is likely not worth refactoring native modules to TypeScript until some of these shortcomings in TypeScript's tooling are addressed, namely, https://github.com/microsoft/TypeScript/issues/21926 and https://github.com/microsoft/TypeScript/issues/420.

**Comment by mchowning** [comment]
> 🔴 Native modules cannot be statically typed checked via CI or editors as there is no way to configure TypeScript to load https://github.com/microsoft/TypeScript/issues/21926 or https://github.com/microsoft/TypeScript/issues/8328#issuecomment-219254627.

It looks like once https://github.com/microsoft/TypeScript/pull/48189 gets released we'll be able to configure TypeScript to load the `.native` files. 😄

**Comment by ciampo** [comment]
**Update**:

a first version of the TypeScript refactoring guidelines has been added to [the `@wordpress/components` package guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript).

From now on, we will be starting to accept help on this task — please follow the guidelines, open a draft PR and either post about it in this issue, or ping me and/or @mirka directly in the PR. 

Thank you!

**Comment by torounit** [comment]
@ciampo @mirka 

Rewrote <FormTokenField> to Typescript.

https://github.com/WordPress/gutenberg/pull/41216

**Comment by t-hamano** [comment]
I have marked `Tooltip` component as a blocker because refactoring is being considered in #42753.

**Comment by kienstra** [comment]
Hope it's OK I added a link to my friend Mike's [ToolbarButton PR](https://github.com/WordPress/gutenberg/pull/47750) to this issue's [top comment](https://github.com/WordPress/gutenberg/issues/35744#issue-1029759945). 

Just wanted to make sure there's no duplication of work.

**Comment by margolisj** [comment]
Looks like `FooterMessageControl` is native only at this point?

**Comment by ciampo** [comment]
> Looks like `FooterMessageControl` is native only at this point?

Good point, I'm going to remove it from the list.

**Comment by mirka** [comment]
I think we can close this tracking issue now that we are practically done! At the component level, we only have the migration to `CustomSelectControl` v2 left, which is progressing and tracked separately (#55023).

Great work everyone 🎉
